csrc/kernel_cache.cpp

```diff
@@ -350,8 +359,64 @@ FusionKernelRuntime* FusionExecutorCache::getKernelRuntimeFor(
     }
   }
 
-  // Access kernels associated with the common device id
-  auto& kernel_runtimes = kernel_runtimes_[args.getDeviceIndex()];
+  auto getConcretizationInfoFromArgs = [args](Fusion* fusion) {
+    ExpressionEvaluator expr_eval;
+
+    // Bind input scalars and tensor metadata to symbolic scalars
+    // Here we bind only the inputs that are needed to concretize dynamic
+    // transforms.
+    TORCH_CHECK(
+        args.size() == fusion->inputs().size(),
+        "Received ",
+        args.size(),
+        " inputs but expected ",
+        fusion->inputs().size());
+    for (auto i : c10::irange(args.size())) {
+      auto inpi = fusion->inputs()[i];
+      if (inpi->isIntegralScalar()) {
+        TORCH_CHECK(
+            args[i]->isType(ArgType::Long),
+            "Expected integer input at position ",
+            i,
+            " but found ",
+            argTypeToString(args[i]->type()));
+
+        const int64_t arg_val =
+            *reinterpret_cast<const int64_t*>(args[i]->arg());
+        expr_eval.bind(inpi, arg_val);
+      } else if (inpi->isA<TensorView>()) {
+        auto tv = inpi->as<TensorView>();
+        auto root = tv->domain()->getRootDomain();
+        TORCH_CHECK(
+            args[i]->isType(ArgType::Tensor),
+            "Expected CUDA tensor at position ",
+            i,
+            " but found ",
+            argTypeToString(args[i]->type()));
+        const TensorArgAbstract* targ =
+            reinterpret_cast<const TensorArgAbstract*>(args[i]);
+        for (auto j : c10::irange(root.size())) {
+          expr_eval.bind(root[j]->extent(), targ->getSize((int64_t)j));
+        }
+      }
+    }
+    return DynamicTransform::getConcretizationInfo(fusion, &expr_eval);
+  };
+
+  // Compute concretization info given inputs. This object points to Vals in
+  // the unconcretized Fusion, so we will not use it directly, but rather it
+  // will be used only as a cache key.
+  std::optional<DynamicTransformConcretizationInfo> conc_info = std::nullopt;
+  if (has_dynamic_reshape_) {
+    conc_info = getConcretizationInfoFromArgs(fusion_.get());
+  }
+
+  // Initialize or fetch vector of FusionKernelRuntime objects associated with
+  // each pair of device ID and
+  auto& kernel_runtimes =
+      kernel_runtimes_
+          .try_emplace(std::make_pair(args.getDeviceIndex(), conc_info), 0)
```

Comment:
What does `std::hash` return for `std::nullopt`? Does it guarantee to return some fixed value?

Reply:
Yeah you bring up a good point. I looked into it before and it seems that it is implementation-specific what value it will return, and whether that value will be the same for `optional<T>(nullopt)` and `optional<U>(nullopt)`. What is guaranteed is that when there's a value the hash will be the same as that of the unwrapped value.  Note that I use a custom hash anyway as we're required to if hashing a `pair`; in an earlier version I had a `pair<size_t, optional<T>>` as the key (also did some spreading of bits) but I changed it to `pair<T, U>` in `SimplePairHash` to simplify it. If we want to guarantee the key for non-symbolic fusions, we could switch back to implementing it for optional.

Reply:
So, for the same device index and there's no concretization info, is it guaranteed to return the same vector?

Reply:
The hash returned for `nullopt` is not specified, but it is required to be the same for equal inputs, so yes it should always give the same vector for same device with `nullopt`.